app/scripts/modules/core/src/search/global/GlobalSearch.tsx

```diff
@@ -173,92 +214,120 @@ export class GlobalSearch extends React.Component<{}, IGlobalSearchState> {
 
   @Debounce(300)
   private executeQuery() {
-    const { infrastructureSearchService } = ReactInjector;
-    const search = infrastructureSearchService.getSearcher();
-
-    ReactGA.event({ category: 'Global Search', action: 'Query', label: this.state.query });
-    this.setState({ querying: true, showMinLengthWarning: false });
-
-    search.query(this.state.query).then((result) => {
-      const showMinLengthWarning = !!this.state.query && this.state.query.length < 3;
-
-      const categories = result
-        .filter(({ results }) => results.length)
-        .map((category) => ({
-          ...category,
-          results: searchRank(category.results).slice(0, 5)
-        }))
-        .sort((a, b) => a.type.order - b.type.order);
+    const { query } = this.state;
 
-      this.resultRefs = range(categories.length).map(() => []);
+    ReactGA.event({ category: 'Global Search', action: 'Query', label: query });
+    this.setState({ querying: true });
 
-      this.setState({
-        querying: false,
-        showSearchResults: !!this.state.query && !showMinLengthWarning,
-        categories,
-        showMinLengthWarning
-      });
-    });
+    this.query$.next(query);
   }
 
+  // Rather than add a jarring warning message as someone is typing a query —
+  // for which a warning might not even be necessary — we wait until typing has finished
+  // to determine whether a warning will be useful.
   @Debounce(300)
   private considerMinLengthWarning() {
     const { query } = this.state;
-    this.setState({ showMinLengthWarning: !!query && query.length < 3 });
-  }
-
-  private showResults() {
-    const { query } = this.state;
-    if (query) {
-      const showMinLengthWarning = query.length < 3;
-      this.setState({
-        showSearchResults: !showMinLengthWarning,
-        showRecentItems: false,
-        showMinLengthWarning
-      });
-    } else {
-      this.showRecentItems();
-    }
+    this.setState({ showMinLengthWarning: !!query && query.length < MIN_SEARCH_LENGTH });
```

Comment:
note:  omitting the `!!` does have slightly different semantics.  If you have a `null`/`undefined` object and you && it, the result is `null`/`undefined` instead of `false`

```
query = null;
this.setState({ showMinLengthWarning: query && query.length < MIN_SEARCH_LENGTH });
// { showMinLengthWarning: null }
```

```
query = null;
this.setState({ showMinLengthWarning: !!query && query.length < MIN_SEARCH_LENGTH });
// { showMinLengthWarning: false }
```

Reply:
Oh interesting, I didn't realize that.

Reply:
Yeah this was very much intentional. In this particular component the `query` is always a string, so `showMinLengthWarning` would end up as `''` in many cases. That case doesn't currently cause any actual change in behavior, but in general I try to keep things type-stable to make reasoning about changes easier later.

The really fun case one often runs into with React where this actually has a UI-visible consequence is the common pattern of putting conditional logic inside a `{}` JSX expression and using something that ends up being non-null:

```jsx
const listOfThings = [];
const maybeFalseyString = '';

return (
  <div>
    {
      // Because listOfThings.length is 0 and not null/undefined/false
      // this renders a text node with the text '0' into the DOM.
      listOfThings.length &&
      <span>There are things in the list of things!</span>
    }
    {
      // Same case as .length, but not particularly disruptive to the DOM.
      // Just adds an empty text node where there should be nothing.
      maybeFalseyString &&
      <span>The string is not empty!</span>
    }
  </div>
);
```

Reply:
Yep, we've run into that several times!